balancer/leastrequest/leastrequest.go

```diff
@@ -0,0 +1,182 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package leastrequest
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/internal/grpcrand"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+// Global to stub out in tests.
+var grpcranduint32 = grpcrand.Uint32
+
+// Name is the name of the least request balancer.
+const Name = "least_request_experimental"
+
+var logger = grpclog.Component("least request")
+
+func init() {
+	balancer.Register(bb{})
+}
+
+// LBConfig is the balancer config for least_request_experimental balancer.
+type LBConfig struct {
+	serviceconfig.LoadBalancingConfig `json:"-"`
+
+	// ChoiceCount is the number of random SubConns to sample to try and find
+	// the one with the Least Request. If unset, defaults to 2. If set to < 2,
+	// will become 2, and if set to > 10, will become 10.
+	ChoiceCount uint32 `json:"choiceCount,omitempty"`
+}
+
+type bb struct{}
+
+func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	lbConfig := &LBConfig{
+		ChoiceCount: 2,
+	}
+	if err := json.Unmarshal(s, lbConfig); err != nil {
+		return nil, fmt.Errorf("least-request: unable to unmarshal LBConfig: %v", err)
+	}
+	// "If `choice_count < 2`, the config will be rejected." - A48
+	if lbConfig.ChoiceCount < 2 { // sweet
+		return nil, fmt.Errorf("least-request: lbConfig.choiceCount: %v, must be >= 2", lbConfig.ChoiceCount)
+	}
+	// "If a LeastRequestLoadBalancingConfig with a choice_count > 10 is
+	// received, the least_request_experimental policy will set choice_count =
+	// 10." - A48
+	if lbConfig.ChoiceCount > 10 {
+		lbConfig.ChoiceCount = 10
+	}
+	return lbConfig, nil
+}
+
+func (bb) Name() string {
+	return Name
+}
+
+func (bb) Build(cc balancer.ClientConn, bOpts balancer.BuildOptions) balancer.Balancer {
+	b := &leastRequestBalancer{scRPCCounts: make(map[balancer.SubConn]*int32)}
+	baseBuilder := base.NewBalancerBuilder(Name, b, base.Config{HealthCheck: true})
+	baseBalancer := baseBuilder.Build(cc, bOpts)
+	b.Balancer = baseBalancer
+	return b
+}
+
+type leastRequestBalancer struct {
+	// Embeds balancer.Balancer because needs to intercept UpdateClientConnState
+	// to learn about choiceCount.
+	balancer.Balancer
```

Comment:
I guess I don't like the base balancer to begin with, so I shouldn't suggest this, but... Is there some way to support configs with it more directly vs. needing this pattern of wrapping it in another balancer to do so?

Reply:
Ummmm so give it a knob like the knob to it to build pickers? I don't like the suggestion either, but let's video call about this if you feel strongly. This feels like a clean way to reuse logic?

Reply:
So I hadn't really thought this through fully, but I think this could work and possibly make things simpler and not be all that difficult to do, either:

- `base.baseBuilder` implements `balancer.ConfigParser`.
- In `baseBuilder.ParseConfig()`, if the associated `base.PickerBuilder` implements `ConfigParser`, then its method is called; otherwise it is a nop.
- Add the parsed LB config (via `UpdateClientConnState()`) to `base.PickerBuildInfo` so that `PickerBuilder`'s `Build()` will have access to it.

This way, you won't need to instantiate a `PickerBuilder`, create a `baseBalancer` with a reference to that, then store (embed) the `baseBalancer` inside the `PickerBuilder`.  Right now there is a cycle (A->B->A), which generally makes things harder to reason about.

Reply:
Discussed offline; decided to keep due to this line in the gRFC: "Counters will be local to each instance of the least_request_experimental policy and thereby not shared in cases where subchannels are used across multiple LB instances." Thus, you need heap construction for a new object that instances this state, and can't use a global picker builder registered in registry.